### PR TITLE
Replace `Thread.sleep(..)` with `Future.get(timeout, :TimeUnit)` in `ClusterCommandExecutor.collectResults(..)`

### LINKED ISSUE #2518: Remove a unnecessary Thread.sleep() from ClusterCommandExecutor.collectResults()
https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/ClusterCommandExecutor.java#L224

when for loop ends with all futures done or cancelled in `ClusterCommandExecutor.collectResults()`.

https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/ClusterCommandExecutor.java#L258

after that, there is no need for thread to be slept, but `Thread.sleep()` is called unnecessarily.

it could be better to check if the `done` is false.

```
try {
    if (!done) {
        Thread.sleep(10);
    }
} catch (InterruptedException e) {
    // ...   
}
```

### PR DESCRIPTION
**Pull request by jxblum**
Resolves #2518 

OTHER NOTABLE CHANGES

Additionally, this PR fixes a couple bugs in `collectResults(..)`, which were called out and captured in [2 test cases](https://github.com/spring-projects/spring-data-redis/blob/issue/2518/src/test/java/org/springframework/data/redis/connection/ClusterCommandExecutorUnitTests.java#L413-L467).

Previously, `ClusterCommandExecutor.collectResults(..)` had no test coverage.

### CODE REVIEW

File: src/main/java/org/springframework/data/redis/connection/ClusterTopology.java [hunk 1]
```diff
@@ -79,13 +81,9 @@ public Set<RedisClusterNode> getActiveNodes() {
...
 	public Set<RedisClusterNode> getActiveMasterNodes() {
 
-		Set<RedisClusterNode> activeMasterNodes = new LinkedHashSet<>(nodes.size());
-		for (RedisClusterNode node : nodes) {
-			if (node.isMaster() && node.isConnected() && !node.isMarkedAsFail()) {
-				activeMasterNodes.add(node);
-			}
-		}
-		return activeMasterNodes;
+		return getNodes().stream()
```
Comment (mp911de) [inline src/main/java/org/springframework/data/redis/connection/ClusterTopology.java:84]: Stream usage introduces a huge performance penalty due to its excessive object allocation. We avoid stream usage to not accidentally introduce a performance regression unless code paths are guaranteed to run once or during the initialization phase.

**Review by mp911de — CHANGES_REQUESTED** [review]
This pull request is a no-go and impossible to review. There are changes all over the place without the possibility to distinguish relevant changes from polishing.

It is fine to address bugs while working on code, but these musts be at least encapsulated in its own commit. 

Also, a lot of these changes create the impression that they are done just because of a different code style and not an improvement over the existing code.

**Comment by jxblum** [comment]
I am not even sure how to respond to this feedback other than not all PRs can be cleanly reviewed through the GitHub [lens](https://github.com/spring-projects/spring-data-redis/pull/2719/files), and sometimes should be viewed as the final product, compared to the original, in your IDE.

The underlying problem from Issue #2518 boils down to this: [before](https://github.com/spring-projects/spring-data-redis/blob/3.1.4/src/main/java/org/springframework/data/redis/connection/ClusterCommandExecutor.java#L220-L263) and [after](https://github.com/spring-projects/spring-data-redis/blob/issue/2518/src/main/java/org/springframework/data/redis/connection/ClusterCommandExecutor.java#L257-L291) Along the way, I refactored other, redundant but related parts of the codebase to improve readability and maintainability.

You cannot properly address the unnecessary and inappropriate `Thread.sleep(..)` invocation without structural and logical changes to the code. This led me to writing tests in the first place since there were no individual unit tests for the `ClusterCommandExecutor.collectResults(..)` method. Through analyzing and testing the code, I also discovered other bugs that I provided test cases for. While edge cases, they were bugs none-the-less.

These changes have very little to do with style vs. correctness, readability and overall long-term maintenance. I feel it is important to address issues the moment we see them and not put them off for later, or be so fine-grained it becomes an impediment to productivity, or even contributions.

In summary, and in total, there were not a lot of wide sweeping changes in this PR, especially when you consider the final product. The goal is always to leave it better than I found it. You can tell me if I am wrong, but I believe I accomplished that.

File: src/main/java/org/springframework/data/redis/connection/ClusterTopology.java (hunk 1 above)
Comment (jxblum) [reply src/main/java/org/springframework/data/redis/connection/ClusterTopology.java:84]: Agreed

Comment (jxblum) [reply src/main/java/org/springframework/data/redis/connection/ClusterTopology.java]: Reverted, completely.

**Comment by jxblum** [comment]
I reworked this PR.

1) First, I rebased this PR on the latest changes/checkins on `main`.
2) Second, I completely removed all modifications to `ClusterTopology` (e.g. removing use of `Streams` among other things).
3) Finally, I reorganized the `ClusterCommandExecutor` logic around `collectResults(..)` even further along with a few other code improvements.

This brings the change set to: `ClusterCommandExecutor`, `ClusterCommandExecutorUnitTests` and `MockitoUtils`. `MockitoUtils` is purely a test utility to facilitate testing with mocks.

**Comment by mp911de** [comment]
That's merged and polished now.
